client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/XaModeTransactionTable.java

```diff
@@ -173,25 +190,50 @@ public void forget(Xid xid) {
          if (trace) {
             log.tracef("XaResource.forget(%s)", xid);
          }
-         //no-op
+         recoveryManager.forgetTransaction(xid);
+         getAnyTransactionContext().forget(xid);
       }
 
       @Override
-      public Xid[] recover(int i) {
-         //no recovery yet
-         return NO_XIDS;
+      public Xid[] recover(int flags) throws XAException {
+         if (trace) {
+            log.tracef("XaResource.recover(%s)", flags);
+         }
+         RecoveryIterator it = this.iterator;
+         if ((flags & XAResource.TMSTARTRSCAN) != 0) {
+            if (it == null) {
+               it = recoveryManager.startScan(getAnyTransactionContext().fetchPreparedTransactions());
+               iterator = it;
+            } else {
+               //we have an iteration in progress.
+               throw new XAException(XAException.XAER_INVAL);
+            }
+         }
+         if ((flags & XAResource.TMENDRSCAN) != 0) {
+            if (it == null) {
+               //we don't have an iteration in progress
+               throw new XAException(XAException.XAER_INVAL);
+            } else {
+               iterator.finish(timeoutMs);
+               iterator = null;
+            }
+         }
+         if (it == null) {
+            //we don't have an iteration in progress
+            throw new XAException(XAException.XAER_INVAL);
+         }
+         return it.next();
       }
 
       @Override
-      public boolean setTransactionTimeout(int i) {
-         //not supported yet
-         return false;
+      public boolean setTransactionTimeout(int timeout) { //timeout is in seconds
```

Comment:
Please move the comment to the next line, or even better rename the parameter to `timeoutSeconds`